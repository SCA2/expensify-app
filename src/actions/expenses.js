import database from '../firebase/firebase';

export const addExpense = (expense) => ({
  type: 'ADD_EXPENSE',
  expense
});

export const startAddExpense = (expenseData = {}) => {
  return (dispatch, getState) => {
    const {
      description = '',
      note = '',
      amount = 0,
      createdAt = 0
    } = expenseData;

    const expense = { description, note, amount, createdAt };
    const path = `users/${getState().auth.uid}/expenses`;

    return database.ref(path).push(expense).then((ref) => {
      dispatch(addExpense({
        id: ref.key,
        ...expense
      }));
    });
  };
};

export const removeExpense = ({ id }) => ({
  type: 'REMOVE_EXPENSE',
  id
});

export const startRemoveExpense = ({ id } = {}) => {
  return (dispatch, getState) => {
    const path = `users/${getState().auth.uid}/expenses/${id}`;
    return database.ref(path).remove().then(() => {
      dispatch(removeExpense({ id }));
    })
  }
};

export const editExpense = (id, updates) => ({
  type: 'EDIT_EXPENSE',
  id,
  updates
});

export const startEditExpense = (id, updates) => {
  return (dispatch, getState) => {
    const path = `users/${getState().auth.uid}/expenses/${id}`;
    return database.ref(path).update(updates).then(() => {
      dispatch(editExpense(id, updates));
    });
  };
};

export const setExpenses = (expenses) => ({
  type: 'SET_EXPENSES',
  expenses
});

export const startSetExpenses = () => {
  return (dispatch, getState) => {
    const path = `users/${getState().auth.uid}/expenses`;
    return database.ref(path).once('value').then((snapshot) => {
      const expenses = [];
      snapshot.forEach(childSnapshot => {
        expenses.push({
          id: childSnapshot.key,
          ...childSnapshot.val()
        });
      });

      dispatch(setExpenses(expenses));
    });
  };
};