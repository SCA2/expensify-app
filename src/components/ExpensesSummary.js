import React from 'react';
import { connect } from 'react-redux';
import numeral from 'numeral';
import selectExpenses from '../selectors/expenses';
import selectExpensesTotal from '../selectors/expenses-total';

export const ExpensesSummary = (props) => (
  <div>
    {`Viewing ${props.expenses.length} expenses totalling ${numeral(props.expensesTotal / 100).format('$0,0.00')}`}
  </div>
);

const mapStateToProps = state => ({
  expenses: selectExpenses(state.expenses, state.filters),
  expensesTotal: selectExpensesTotal(state.expenses)
});

export default connect(mapStateToProps)(ExpensesSummary);