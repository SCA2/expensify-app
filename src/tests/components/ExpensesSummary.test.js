import React from 'react';
import { shallow } from 'enzyme';
import { ExpensesSummary } from '../../components/ExpensesSummary';
import { expenses } from '../fixtures/expenses';
import selectExpensesTotal from '../../selectors/expenses-total';

test('can display total of one expense', () => {
  const wrapper = shallow(<ExpensesSummary
    expenses={[expenses[0]]}
    expensesTotal={selectExpensesTotal([expenses[0]])}
  />);
  expect(wrapper).toMatchSnapshot();
});

test('can display total of multiple expenses', () => {
  const wrapper = shallow(<ExpensesSummary
    expenses={expenses}
    expensesTotal={selectExpensesTotal(expenses)}
  />);
  expect(wrapper).toMatchSnapshot();
});