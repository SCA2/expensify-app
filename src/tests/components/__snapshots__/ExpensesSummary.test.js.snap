// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`can display total of multiple expenses 1`] = `
<h1>
  Viewing 3 expenses totalling $1,141.95
</h1>
`;

exports[`can display total of one expense 1`] = `
<h1>
  Viewing 1 expense totalling $1.95
</h1>
`;
